lndr-backend/src/Lndr/Db/VerifiedCredits.hs

```diff
@@ -73,15 +73,19 @@ verifyCreditByHash :: Text -> Connection -> IO Int
 verifyCreditByHash hash conn = fromIntegral <$> execute conn "UPDATE settlements SET verified = TRUE WHERE hash = ?" (Only hash)
 
 
+-- Balance is calculated per ucac and only includes bilateral non-settlement
+-- credits and transaction-verified bilateral settlementcredits
 userBalance :: Address -> Address -> Connection -> IO Integer
 userBalance addr ucac conn = do
-    [Only balance] <- query conn "SELECT (SELECT COALESCE(SUM(amount), 0) FROM verified_credits WHERE creditor = ? AND ucac = ?) - (SELECT COALESCE(SUM(amount), 0) FROM verified_credits WHERE debtor = ? AND ucac = ?)" (addr, ucac, addr, ucac) :: IO [Only Scientific]
+    [Only balance] <- query conn "SELECT (SELECT COALESCE(SUM(verified_credits.amount), 0) FROM verified_credits LEFT JOIN settlements USING(hash) WHERE creditor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE) - (SELECT COALESCE(SUM(verified_credits.amount), 0) FROM verified_credits LEFT JOIN settlements USING (hash) WHERE debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE)" (addr, ucac, addr, ucac) :: IO [Only Scientific]
     return . floor $ balance
 
 
+-- Two-party balance is calculated per ucac and only includes bilateral
+-- non-settlement credits and transaction-verified bilateral settlementcredits
 twoPartyBalance :: Address -> Address -> Address -> Connection -> IO Integer
 twoPartyBalance addr counterparty ucac conn = do
-    [Only balance] <- query conn "SELECT (SELECT COALESCE(SUM(amount), 0) FROM verified_credits WHERE creditor = ? AND debtor = ? AND ucac = ?) - (SELECT COALESCE(SUM(amount), 0) FROM verified_credits WHERE creditor = ? AND debtor = ? AND ucac = ?)" (addr, counterparty, ucac, counterparty, addr, ucac) :: IO [Only Scientific]
+    [Only balance] <- query conn "SELECT (SELECT COALESCE(SUM(verified_credits.amount), 0) FROM verified_credits LEFT JOIN settlements USING(hash) WHERE creditor = ? AND debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE) - (SELECT COALESCE(SUM(verified_credits.amount), 0) FROM verified_credits LEFT JOIN settlements USING(hash) WHERE creditor = ? AND debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE)" (addr, counterparty, ucac, counterparty, addr, ucac) :: IO [Only Scientific]
```

Comment:
First pass:

```sql
SELECT (
  SELECT
    COALESCE(SUM(verified_credits.amount), 0)
  FROM
    verified_credits LEFT JOIN settlements USING(hash)
  WHERE
    creditor = ? AND debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE
) - (
  SELECT
    COALESCE(SUM(verified_credits.amount), 0)
  FROM
    verified_credits LEFT JOIN settlements USING(hash)
  WHERE
    creditor = ? AND debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE
)
```

Second pass:

```sql
SELECT
  COALESCE(SUM(verified_credits.amount), 0)
FROM
  verified_credits LEFT JOIN settlements USING(hash)
WHERE
  creditor = ? AND debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE
```

…and…

```sql
SELECT
  COALESCE(SUM(verified_credits.amount), 0)
FROM
  verified_credits LEFT JOIN settlements USING(hash)
WHERE
  creditor = ? AND debtor = ? AND ucac = ? AND verified IS DISTINCT FROM FALSE
```

Similar queries here to `userBalance`, except in this case, the two individual SQL queries are **identical**! Same advice goes for this as for above, then, except here we only need to decompose into one additional function, to calculate the total volume of credit from one user to another. We'd just call that once for one direction, again for the other direction, and in Haskell land, take the difference between the two results.

Reply:
Yeah, I don't think it's worth using two round trips to the DB, but I know this looks better. Originally, the code was exactly as you described.